src/transformStateWithClones.js

```diff
@@ -7,7 +7,29 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const result = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(currentState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete currentState[key];
```

Comment:
Directly deleting properties from currentState can be avoided by making a copy first and then deleting from the copy. This aligns with the checklist item: 'don't mutate object or arrays - it will cause unexpected results later on. You should make copy using Object.assign or spread operator.'